src/web/ng/impl/ConnectionHandler.cpp

```diff
@@ -0,0 +1,285 @@
+{
+    onStop_();
+}
+
+bool
+ConnectionHandler::handleError(Error const& error, Connection const& connection) const
+{
+    // ssl::error::stream_truncated, also known as an SSL "short read",
+    // indicates the peer closed the connection without performing the
+    // required closing handshake (for example, Google does this to
+    // improve performance). Generally this can be a security issue,
+    // but if your communication protocol is self-terminated (as
+    // it is with both HTTP and WebSocket) then you may simply
+    // ignore the lack of close_notify.
+    //
+    // https://github.com/boostorg/beast/issues/38
+    //
+    // https://security.stackexchange.com/questions/91435/how-to-handle-a-malicious-ssl-tls-shutdown
+    //
+    // When a short read would cut off the end of an HTTP message,
+    // Beast returns the error boost::beast::http::error::partial_message.
+    // Therefore, if we see a short read here, it has occurred
+    // after the message has been completed, so it is safe to ignore it.
+    if (error == boost::beast::http::error::end_of_stream || error == boost::asio::ssl::error::stream_truncated)
+        return false;
+
+    // WebSocket connection was gracefully closed
+    if (error == boost::beast::websocket::error::closed)
+        return false;
+
+    if (error != boost::asio::error::operation_aborted) {
+        LOG(log_.error()) << connection.tag() << ": " << error.message() << ": " << error.value();
+    }
+    return true;
+}
+
+bool
+ConnectionHandler::sequentRequestResponseLoop(Connection& connection, boost::asio::yield_context yield)
+{
+    // The loop here is infinite because:
+    // - For websocket connection is persistent so Clio will try to read and respond infinite unless client
+    //   disconnected.
+    // - When client disconnected connection.send() or connection.receive() will return an error.
+    // - For http it is still a loop to reuse the connection if keep alive is set. Otherwise client will disconnect and
+    //   an error appears.
+    // - When server is shutting down it will cancel all operations on the connection so an error appears.
+
+    while (true) {
+        auto expectedRequest = connection.receive(yield);
+        if (not expectedRequest)
+            return handleError(expectedRequest.error(), connection);
+
+        LOG(log_.info()) << connection.tag() << "Received request from ip = " << connection.ip();
+
+        auto maybeReturnValue = processRequest(connection, std::move(expectedRequest).value(), yield);
+        if (maybeReturnValue.has_value())
+            return maybeReturnValue.value();
+    }
+}
+
+bool
+ConnectionHandler::parallelRequestResponseLoop(Connection& connection, boost::asio::yield_context yield)
+{
+    // atomic_bool is not needed here because everything happening on coroutine's strand
+    bool stop = false;
+    bool closeConnectionGracefully = true;
+    util::CoroutineGroup tasksGroup{yield, maxParallelRequests_};
```

Comment:
Maybe the limit of group size should be got from DosGuard

Reply:
Webserver should be completely independent from any other functionality. Also dosguard will do that just on a higher level. This is a protection to not spawn too many coroutines.